flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamDataStreamPythonFunctionRunner.java

```diff
@@ -102,85 +106,144 @@ public BeamDataStreamPythonFunctionRunner(
 
     @Override
     protected void buildTransforms(RunnerApi.Components.Builder componentsBuilder) {
-        for (int i = 0; i < userDefinedDataStreamFunctions.size() + 1; i++) {
-            String functionUrn;
-            if (i == 0) {
-                functionUrn = headOperatorFunctionUrn;
-            } else {
-                functionUrn = STATELESS_FUNCTION_URN;
-            }
+        for (int i = 0; i < userDefinedDataStreamFunctions.size(); i++) {
 
-            FlinkFnApi.UserDefinedDataStreamFunction functionProto;
-            if (i < userDefinedDataStreamFunctions.size()) {
-                functionProto = userDefinedDataStreamFunctions.get(i);
-            } else {
-                // the last function in the operation tree is used to prune the watermark column
-                functionProto = createReviseOutputDataStreamFunctionProto();
-            }
+            final Map<String, String> outputCollectionMap = new HashMap<>();
 
-            // Use ParDoPayload as a wrapper of the actual payload as timer is only supported in
-            // ParDo
-            final RunnerApi.ParDoPayload.Builder payloadBuilder =
-                    RunnerApi.ParDoPayload.newBuilder()
-                            .setDoFn(
-                                    RunnerApi.FunctionSpec.newBuilder()
-                                            .setUrn(functionUrn)
-                                            .setPayload(
-                                                    org.apache.beam.vendor.grpc.v1p26p0.com.google
-                                                            .protobuf.ByteString.copyFrom(
-                                                            functionProto.toByteArray()))
-                                            .build());
-
-            // Timer is only available in the head operator
-            if (i == 0 && timerCoderDescriptor != null) {
-                payloadBuilder.putTimerFamilySpecs(
-                        TIMER_ID,
-                        RunnerApi.TimerFamilySpec.newBuilder()
-                                // this field is not used, always set it as event time
-                                .setTimeDomain(RunnerApi.TimeDomain.Enum.EVENT_TIME)
-                                .setTimerFamilyCoderId(WRAPPER_TIMER_CODER_ID)
-                                .build());
+            // Prepare side outputs
+            if (i == userDefinedDataStreamFunctions.size() - 1) {
```

Comment:
Does it make sense to move this out of the for loop?

Reply:
This add outputs to local variable `outputCollectionMap` in the loop.